File: Data Structures & Algorithms/stone-game/submission-4.cpp
class Solution {
public:
    map <vector<int>, int> memo;
    bool stoneGame(vector<int>& piles) {
        int res = dfs(0, piles.size() - 1, piles);
        int tot = accumulate(piles.begin(), piles.end(), 0);
        return res > tot / 2;
    }
    int dfs(int l, int r, vector<int> &piles) {
        if (l > r) return 0;
        if (memo.count({l, r}) == 1) return memo[{l, l}];
        if ((r - l) % 2 == 0) {
            int left = piles[l] + dfs(l+1, r, piles);
            int right = piles[r] + dfs(l, r-1, piles);
            memo[{l, r}] = max(left, right);
        }
        else {
            memo[{l, r}] = max(dfs(l + 1, r, piles), dfs(l, r-1, piles));
        }
        return memo[{l, r}];
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        res = 0
        def dfs(root, val):
            nonlocal res
            if not root:
                return
            if root.val >= val:
                val = root.val
                res+=1
            dfs(root.left, val)
            dfs(root.right, val)
        dfs(root, root.val)
        return res